### DISCUSSION TITLE
Access Node Database Decision

### DISCUSSION THREAD

**Discussion by oriolcanades** [DOME-Marketplace/architecture#2] [ADR]
# Access Node Database Decision

## Status

The current status of this decision is **PROPOSED**.

## Context

The current Access Node instance three databases. A Redis database for the TM Forum API component (if needed), a PostgreSQL for the Context Broker and a PostgreSQL for the Replication Service. Actually, the team in charge of developing the fourth component of the Access Node wants to implement its own database. If we consider the current architecture, we will have four different databases in the Access Node. This could lead to a more complex architecture and more difficult to maintain because each one implements a single instance of a database.

```mermaid
graph TB
    marketplace[Marketplace]
    subgraph "Access Node"
        tm_forum_api[TM Forum API]
        redis[(Redis)]
        context_broker[Context Broker]
        postgis[(PostgreSQL)]
        replication_service[Replication Service]
        postgres[(PostgreSQL)]
        dlt_adapter[DLT Adapter]
    end
    blockchain[Blockchain]
    marketplace --> tm_forum_api
    tm_forum_api --> redis
    context_broker --> postgis
    replication_service --> postgres
    tm_forum_api --> context_broker
    context_broker --> replication_service
    replication_service --> dlt_adapter
    
    dlt_adapter --> blockchain
```

## Decision

We propose an approach to using a single database for all components of the Access Node. This approach will simplify the architecture and make it easier to maintain. Nevertheless, we need to consider that each component will need to implement its own schema in the database separating the responsibility of each component.

```mermaid
graph TB
    marketplace[Marketplace]
    subgraph "Access Node"
        tm_forum_api[TM Forum API]
        context_broker[Context Broker]
        replication_service[Replication Service]
        dlt_adapter[DLT Adapter]
        postgres[(PostgreSQL)]
    end
    blockchain[Blockchain]
    marketplace --> tm_forum_api
    tm_forum_api -->|Schema: schema_tm_forum| postgres
    context_broker -->|Schema: schema_context| postgres
    replication_service -->|Schema: schema_replication| postgres
    dlt_adapter -->|Schema: schema_dlt| postgres
    tm_forum_api --> context_broker
    context_broker --> replication_service
    replication_service --> dlt_adapter
    dlt_adapter --> blockchain
```

## Consequences

One of the difficulties could be that some of the current databases implement or have some bound configurations, and we will need to change the static configuration to external configuration. This could lead to a more complex configuration for the database. Nevertheless, we believe that the benefits of having a single database for all components outweigh the drawbacks.

**Comment by hesusruiz** [comment]
The decisions:

It is FORBIDEN to deploy PosgreSQL or any other standard database as part of the Access Node.

There are only two options:
1. Use a DBaaS provided by IONOS
2. Use SQLite for a specific application-database, when the use case required it.

**Comment by oriolcanades** [comment]
**Decision**:  
It is **FORBIDDEN** to deploy PostgreSQL or any other standard database as part of the Access Node.

**Available options**:
1. Use a DBaaS (Database as a Service) provided by IONOS.
2. Use SQLite for specific application-database requirements when the use case necessitates it.

## Database Management for Microservices  
After reviewing the pros and cons of using shared vs. separate databases for microservices, I align with the perspective shared in [this article](https://dev.to/gajus/designing-microservices-with-shared-databases-14kk):

> _"In short, setting up a separate database for a new project always sounds nice in theory, but in practice leads to a ton of overhead down the road. Keep things simple by using one database and give every user narrow permissions to make the changes to the database predictable."_

### Why not separate databases or schemas for each microservice?

- **Loss of ACID transactions**: Separating databases removes the ability to perform ACID-compliant transactions across services.
- **Loss of referential integrity**: Relationships between different parts of the system may not be easily enforceable.
- **Loss of PITR (Point-in-Time Recovery)**: Recovery processes become more complex and less reliable.
- **Increased complexity**: Deployment, permission management, and keeping databases in sync become harder and more error-prone.

### Recommended Approach: Granular Permissions

Since implementing separate databases or schemas for each microservice/system is not ideal, we propose setting **granular database permissions** to ensure each microservice can only access and manipulate its own data while maintaining a shared database structure.

#### Example SQL Approach:

```sql
CREATE USER 'context_broker_service'@'localhost' IDENTIFIED BY 'password1';
CREATE USER 'replication_service'@'localhost' IDENTIFIED BY 'password2';
CREATE USER 'dlt_adapter_service'@'localhost' IDENTIFIED BY 'password3';

GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, DROP ON mydb.context_broker_table TO 'context_broker_service'@'localhost';
GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, DROP ON mydb.replication_table TO 'replication_service'@'localhost';
GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, DROP ON mydb.dlt_adapter_table TO 'dlt_adapter_service'@'localhost';
```

This way, each service (e.g., context broker, replication service, and DLT adapter) will have limited and controlled access to its own subset of the shared database, ensuring security and reducing the risk of side effects between microservices.

@wistefan, is this approach feasible for implementation within the Scorpio Context Broker? If so, it could help streamline database management while ensuring that each microservice operates within a well-defined access perimeter.

## Architectural View
```mermaid
graph TB
    marketplace[Marketplace]
    subgraph "Access Node"
        tm_forum_api[TM Forum API]
        context_broker[Context Broker]
        replication_service[Replication Service]
        dlt_adapter[DLT Adapter]
    end
    subgraph "DBaaS"
        postgres[(PostgreSQL)]
    end
    blockchain[Blockchain]
    marketplace --> tm_forum_api
    context_broker -->|user: context_broker_service | postgres
    replication_service -->|user: replication_service | postgres
    dlt_adapter -->|user: dlt_adapter_service | postgres
    tm_forum_api --> context_broker
    context_broker --> replication_service
    replication_service --> dlt_adapter
    dlt_adapter --> blockchain
```

**Comment by wistefan** [comment]
Scorpio requires a PostgreSQL with the Postgis-Extension enabled. If the IONOS service is capable of providing that, it can be used. 

Some additional notes:
* The TMForum Implementation uses Redis as an in-memory cache, not as a database. Its required for horizontal scaling, else notifications wont properly work. TMForum itself does not have any database.
* The Access-Node is a component that will not only be deployed as part of the main DOME instance, but also by the integrators. They likely are not running in IONOS-infrastructure. Whats the recommended solution for them?
* I do not understand the rule about SQLite for multiple reasons:
  * The single file approach of SQLite is a big issue for cloud-based systems:
    - when the service is rescheduled to another zone of the cluster, the file-location might not be available anymore
    - when scaling horizontally the file must be read-write for multiple instances
    - it couples state to the application, without having it explicitly visible(as it is with an explicit database)
    
    -> properly using SQLite in a cloud-based system requires Perstistent Volumes through a network-filesystem that supports mulit-read/write. Kubernetes is typically run with multi-read/single-write. This increases the operational complexity of the cluster, in my opinion far more than operating a database like PostgreSQL/MariaDB. 
  * Since a Postgres compatible DataBase already has to be part of the AccessNode, why is another SQL option required at all? I would understand alternatives for NoSQL, Timeseries or GraphDBs, if required(which I dont think it is), but everything that can be done with SQLite can already be done with the default SQL-Database
 * Even thought I agree with the single-instance Database solution as a good solution for the default AccessNode, the Schemas(in contrast to the mentioned [shared-database pattern](https://microservices.io/patterns/data/shared-database.html)) need to be seperated. In contrast to the pattern's use-case, there is no direct transactional coupling between the services in the access-node and since the services are developed by different teams and companies, the required coupling of development and runtime is not an option. Components like Scorpio are developed outside of DOME, thus its being impossible to couple them on that level.

 
@oriolcanades  can you answer some of my question? Especially why  it was decided to drop PostgreSQL at all from the AccessNode? This creates a huge vendor-lock to IONOS for all integrators, which is most likely not acceptable.

**Comment by oriolcanades** [comment]
I am not sure if IONOS DBaaS supports the Postgis-Extension. We will need to reach out to them for clarification. 

The decision to remove the database from the Access Node boundary was made to allow both DOME and non-DOME implementers the flexibility to use a database, whether On-Premise or SaaS. If you prefer an On-Premise database, you could deploy it within your Kubernetes cluster and configure the Access Node to use it.

I completely agree that the Access Node will be deployed not only as part of the main DOME instance but also by integrators. We should propose and recommend that implementers can choose a DBaaS that meets the specifications or deploy an On-Premise instance with the Postgis-Extension in a separate POD, outside the conceptual boundaries of the Access Node.

I am not opposed to differentiating the data boundaries using schemas or applying a user role model. My initial plan already included the use of separate schemas ;)

```mermaid
graph TB
    marketplace[Marketplace]
    subgraph "Access Node"
        tm_forum_api[TM Forum API]
        context_broker[Context Broker]
        replication_service[Replication Service]
        dlt_adapter[DLT Adapter]
    end
    subgraph "DBaaS or On-Premise"
        postgres[(PostgreSQL + Postgis Ext)]
    end
    blockchain[Blockchain]
    marketplace --> tm_forum_api
    context_broker -->|Schema: context_broker| postgres
    replication_service -->|Schema: replication_service| postgres
    dlt_adapter -->|Schema: dlt_adapter| postgres
    tm_forum_api --> context_broker
    context_broker --> replication_service
    replication_service --> dlt_adapter
    dlt_adapter --> blockchain
```

**Reply by alejandro-nieto-git** [reply]
I think this is the way to go

**Comment by wistefan** [comment]
@oriolcanades Thanks for the answer. 
>The decision to remove the database from the Access Node boundary was made to allow both DOME and non-DOME implementers the flexibility to use a database, whether On-Premise or SaaS. If you prefer an On-Premise database, you could deploy it within your Kubernetes cluster and configure the Access Node to use it.

In that case, I think it the decision should be better clarified. I agree that the Database can be seen as an outside infrastructure requirement, but than it should not say "Its forbidden to deploy a Postgres", but instead:
The AccessNode requires an PostgreSQL(+Postgis) compatible database. The recommendation is to use a DBaaS if available or <insert alternative solution>(in my opinion, the postgres helm chart;)).
